ui/apps/platform/src/Containers/Policies/Wizard/Step3/PolicyCriteriaModal.tsx

```diff
@@ -50,26 +56,81 @@ function getKeysByCategory(keys) {
     return categories;
 }
 
+function getPolicyFieldsAsTree(existingGroups, descriptors): TreeViewDataItem[] {
+    const categories = getKeysByCategory(descriptors);
+
+    const treeList = Object.keys(categories).map((category) => ({
+        name: '',
+        title: category,
+        id: kebabCase(category),
+        children: categories[category]
+            .filter((child) => {
+                const alreadyUsed = existingGroups.find(
+                    (group) =>
+                        group.fieldName.toLowerCase() === child?.name?.toLowerCase() ||
+                        group.fieldName.toLowerCase() === child?.label?.toLowerCase()
+                );
+                return !alreadyUsed;
+            })
+            .map<TreeViewDataItem>((child: Descriptor) => ({
+                name: child.longName,
+                title: child.shortName || child.name,
+                id: kebabCase(child.shortName),
+            })),
+    }));
+
+    return treeList;
+}
+
 function PolicyCriteriaModal({
     addPolicyFieldCardHandler,
     descriptors,
+    existingGroups,
     isModalOpen,
     onClose,
 }: PolicyCriteriaModalProps) {
     const [activeItems, setActiveItems] = useState<TreeViewDataItem[]>([]);
     const [allExpanded, setAllExpanded] = useState(false);
+    const [filteredItems, setFilteredItems] = useState<TreeViewDataItem[]>([]);
+    const [isFiltered, setIsFiltered] = useState(false);
 
-    const categories = getKeysByCategory(descriptors);
-    const treeList = Object.keys(categories).map((category) => ({
-        name: '',
-        title: category,
-        id: kebabCase(category),
-        children: categories[category].map((child) => ({
-            name: child.longName,
-            title: child.shortName,
-            id: kebabCase(child.shortName),
-        })),
-    }));
+    const treeDataItems = useMemo(
+        () => getPolicyFieldsAsTree(existingGroups, descriptors),
+        [descriptors, existingGroups]
+    );
+
+    useEffect(() => {
+        setFilteredItems(treeDataItems);
+    }, [treeDataItems]);
```

Comment:
Not necessarily requesting a change, but I found it interesting that React now recommends setting state directly during rendering over using an effect to update state after a prop change. https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes

From what I understand, something like below would prevent a full cycle of [update state -> commit to DOM -> use effect -> update state -> commit to DOM].
```typescript
    const [prevTreeItems, setPrevTreeItems] = useState(treeDataItems);

    if (treeDataItems !== prevTreeItems) {
        setPrevTreeItems(treeDataItems);
        setFilteredItems(treeDataItems);
    }
```
This seems a little verbose and error prone though so I imagine if this is something we wanted to do more widespread we would wrap it in a custom hook.

Reply:
If we keep the effect - do we need `setIsFiltered(false);` as well here?